src/java/org/apache/cassandra/nodes/virtual/LegacySystemKeyspaceToNodes.java

```diff
@@ -0,0 +1,293 @@
+ */
+package org.apache.cassandra.nodes.virtual;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.UntypedResultSet;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.SystemKeyspace;
+import org.apache.cassandra.db.commitlog.CommitLogPosition;
+import org.apache.cassandra.db.marshal.BytesType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UUIDType;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.util.DataInputBuffer;
+import org.apache.cassandra.io.util.RebufferingInputStream;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.nodes.BootstrapState;
+import org.apache.cassandra.nodes.LocalInfo;
+import org.apache.cassandra.nodes.NodeInfo;
+import org.apache.cassandra.nodes.Nodes;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.SchemaConstants;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.utils.CassandraVersion;
+
+import static java.lang.String.format;
+import static org.apache.cassandra.cql3.QueryProcessor.executeInternal;
+
+/**
+ * Upgrade code to migrate from {@code system.local} and {@code system.peers} to {@link Nodes}.
+ * This migration exists to enable the upgrade path from older DSE versions that store the local and peers data in
+ * sstables to flat files that are used within CC.
+ * Without it the node isn't able (among others) to read its host id after upgrade.
+ */
+public final class LegacySystemKeyspaceToNodes
+{
+    private static final Logger logger = LoggerFactory.getLogger(LegacySystemKeyspaceToNodes.class);
+
+    @VisibleForTesting
+    static final AtomicBoolean isRunning = new AtomicBoolean(false);
+
+    public int peersMigrated;
+    public boolean localMigrated;
+    private final Nodes nodes;
+
+    public static boolean isRunning()
+    {
+        return isRunning.get();
+    }
+
+    /**
+     * Executes the system.local and system.peers tables migration to flat files
+     * @return true if the flat files didn't exist and the tables contained data to be migrated, false otherwise
+     */
+    public static void convertToNodesFilesIfNecessary()
+    {
+        try
+        {
+            if (Nodes.local().localInfoWasPresent())
+            {
+                return;
+            }
+
+            new LegacySystemKeyspaceToNodes(Nodes.nodes()).convertToNodesFiles();
```

Comment:
trivial and optional…
```suggestion
            if (!Nodes.local().localInfoWasPresent())
                new LegacySystemKeyspaceToNodes(Nodes.nodes()).convertToNodesFiles();
```

Reply:
Fixed.